polkadot/runtime/parachains/src/scheduler.rs

```diff
@@ -531,50 +337,20 @@ impl<T: Config> Pallet<T> {
 	/// Return the next thing that will be scheduled on this core assuming it is currently
 	/// occupied and the candidate occupying it became available.
 	pub(crate) fn next_up_on_available(core: CoreIndex) -> Option<ScheduledCore> {
-		ClaimQueue::<T>::get()
-			.get(&core)
-			.and_then(|a| a.front().map(|pe| Self::paras_entry_to_scheduled_core(pe)))
-	}
-
-	fn paras_entry_to_scheduled_core(pe: &ParasEntryType<T>) -> ScheduledCore {
-		ScheduledCore { para_id: pe.para_id(), collator: None }
-	}
-
-	/// Return the next thing that will be scheduled on this core assuming it is currently
-	/// occupied and the candidate occupying it times out.
-	pub(crate) fn next_up_on_time_out(core: CoreIndex) -> Option<ScheduledCore> {
-		let max_availability_timeouts = configuration::ActiveConfig::<T>::get()
-			.scheduler_params
-			.max_availability_timeouts;
-		Self::next_up_on_available(core).or_else(|| {
-			// Or, if none, the claim currently occupying the core,
-			// as it would be put back on the queue after timing out if number of retries is not at
-			// the maximum.
-			let cores = AvailabilityCores::<T>::get();
-			cores.get(core.0 as usize).and_then(|c| match c {
-				CoreOccupied::Free => None,
-				CoreOccupied::Paras(pe) =>
-					if pe.availability_timeouts < max_availability_timeouts {
-						Some(Self::paras_entry_to_scheduled_core(pe))
-					} else {
-						None
-					},
-			})
+		ClaimQueue::<T>::get().get(&core).and_then(|a| {
+			a.front()
+				.map(|assignment| ScheduledCore { para_id: assignment.para_id(), collator: None })
 		})
 	}
 
 	/// Pushes occupied cores to the assignment provider.
-	fn push_occupied_cores_to_assignment_provider() {
-		AvailabilityCores::<T>::mutate(|cores| {
-			for core in cores.iter_mut() {
-				match core::mem::replace(core, CoreOccupied::Free) {
-					CoreOccupied::Free => continue,
-					CoreOccupied::Paras(entry) => {
-						Self::maybe_push_assignment(entry);
-					},
-				}
-			}
-		});
+	fn push_occupied_cores_to_assignment_provider(
+		occupied_cores: impl Iterator<Item = (CoreIndex, ParaId)>,
+	) {
+		for (core_index, para_id) in occupied_cores {
+			// TODO: can we assume a pool assignment for simplicity?
```

Comment:
I removed the `AvailabilityCores` storage, so we iterate through the `PendingAvailability` storage of the inclusion module to get the occupied cores.
Now, this storage does not hold data about the assignment type (Bulk or Pool), only knows the paraid.

We could add this to the storage but I would want to avoid that if possible. @eskimor you know the most about the coretime assigner logic.
What would happen if the parachain was scheduled on a core in Bulk and we'd push a Pool assignment to that core?

Reply:
Let's just change the logic. We know ahead of time that the next block will be in a new session: Instead of occupying the core and running availability, despite the fact that we already know it won't make it ... we could just not occupy the core and drop the candidate, pushing back the claim to the assignment provider. 

For candidates that are already occupying for longer ... just drop them, they had their chance.